# Import PANDAS Library

Notice that we are using pd as an abbreviation so each time we want to use the PANDAS library, we can simply type pd instead of pandas.

In [190]:
import pandas as pd
pd.options.display.float_format = "{:,.2f}".format

# Import Dataset (either full dataset or simplified)

You can use either set of datafiles (full or simplified). In the past some students' computers could not handle the full dataset. If you worry that your computer will not handle the full dataset (or if you find your code is running very slow), use the simplified version. 

The simplified version was created by taking a sample of the full data; hence the simplified files have "Sample" added to the end of the file names. Some files were not large to begin with and didn't need to be reduced. Those files do not have Sample added to the end of the file name.

## Adjust the file address below to match your computer.

You will want to replace the begining of the address below (i.e., everything before OneDrive - Oregon State University/)

begInv = pd.read_csv("C:/Users/peschh/OneDrive - Oregon State University/ACTG379_DataShare/Bibitor_Data/2016Data/SimplifiedData/HW4 Sampled Data/BegInvFINAL12312016Sample.csv")

## Deliverable 1A

In [191]:
begInv = pd.read_csv("BegInvFINAL12-31-16Sample.csv")

In [192]:
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2016-01-01  
1  750mL       7  10.99  2016-01-01

In [193]:
begInv = begInv.rename(columns = {'onHand':'01_01_2016_Qty'})
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  01_01_2016_Qty  Price   startDate  
0  750mL               8  12.99  2016-01-01  
1  750mL               7  10.99  2016-01-01

In [194]:
begInv['InvCost01_01_2016'] = begInv['01_01_2016_Qty'] * begInv['Price']
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  01_01_2016_Qty  Price   startDate  InvCost01_01_2016  
0  750mL               8  12.99  2016-01-01             103.92  
1  750mL               7  10.99  2016-01-01              76.93

In [195]:
begInv_Store = begInv[['Store','01_01_2016_Qty','InvCost01_01_2016']].groupby('Store').sum()
begInv_Store.head(10)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
1               49917         838,079.59
2               52925         840,452.20
3                8379         132,241.23
4               56819         786,488.68
5               17968         218,720.64
6               48512         739,355.88
7               48330         775,694.77
8               38803         554,182.37
9               52169         713,163.99
10              84728       1,350,761.19

## Deliverable 1B

In [196]:
begInv_Store.sort_values(by='InvCost01_01_2016',ascending=False).head(10)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
34             153852       3,291,170.24
73             162551       3,142,497.36
67             158996       3,079,578.63
66             149314       2,973,033.90
76             140208       2,952,418.44
69             144255       2,946,726.65
38             114368       2,232,698.77
55             119641       2,001,263.66
50              94720       1,649,808.22
79              95330       1,503,149.48

In [197]:
begInv_Store.sort_values(by='InvCost01_01_2016').tail(10).sort_values(by='InvCost01_01_2016', ascending = False)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
34             153852       3,291,170.24
73             162551       3,142,497.36
67             158996       3,079,578.63
66             149314       2,973,033.90
76             140208       2,952,418.44
69             144255       2,946,726.65
38             114368       2,232,698.77
55             119641       2,001,263.66
50              94720       1,649,808.22
79              95330       1,503,149.48

## Deliverable 2A

In [198]:
endInv = pd.read_csv("EndInvFINAL12-31-16Sample.csv")

In [199]:
endInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2016-12-31  
1  750mL       7  36.99  2016-12-31

In [200]:
endInv = endInv.rename(columns = {'onHand':'12_31_2016_Qty'}) 

In [201]:
endInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  12_31_2016_Qty  Price     endDate  
0  750mL              11  12.99  2016-12-31  
1  750mL               7  36.99  2016-12-31

In [202]:
endInv['InvCost12_31_2016'] = endInv['12_31_2016_Qty'] * endInv['Price']
endInv.head(2) 

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  12_31_2016_Qty  Price     endDate  InvCost12_31_2016  
0  750mL              11  12.99  2016-12-31             142.89  
1  750mL               7  36.99  2016-12-31             258.93

In [203]:
endInv_Store.sort_values(by='InvCost12_31_2016').tail(10).sort_values(by='InvCost12_31_2016', ascending = False)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
50             260717       4,887,260.68
73             164589       3,254,662.81
67             163765       3,076,114.82
34             145829       3,074,616.75
76             143866       2,975,945.18
69             150848       2,968,678.82
66             144579       2,860,504.99
74             166015       2,803,645.13
38             129397       2,463,906.85
55             125584       2,234,836.35

## Deliverable 2B

In [204]:
endInv_Store.sort_values(by='InvCost12_31_2016',ascending=True).head(10)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
46                  0               0.00
3                8470         125,674.17
29              21341         236,009.36
65              19663         247,155.74
5               21708         255,093.41
17              20521         260,691.26
45              22083         283,962.09
18              21012         284,557.99
43              22402         289,753.36
16              25198         316,166.19

In [205]:
endInv_Store.sort_values(by='InvCost12_31_2016').tail(10).sort_values(by='InvCost12_31_2016', ascending = True)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
55             125584       2,234,836.35
38             129397       2,463,906.85
74             166015       2,803,645.13
66             144579       2,860,504.99
69             150848       2,968,678.82
76             143866       2,975,945.18
34             145829       3,074,616.75
67             163765       3,076,114.82
73             164589       3,254,662.81
50             260717       4,887,260.68

## Deliverable 3

In [206]:
begInv.head(1) 

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   

    Size  01_01_2016_Qty  Price   startDate  InvCost01_01_2016  
0  750mL               8  12.99  2016-01-01             103.92

In [207]:
endInv.head(1)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   

    Size  12_31_2016_Qty  Price     endDate  InvCost12_31_2016  
0  750mL              11  12.99  2016-12-31             142.89

In [208]:
begInv = begInv.rename(columns = {'01_01_2016_Qty':'BeginningQuantity'})

In [209]:
endInv = endInv.rename(columns = {'12_31_2016_Qty':'EndingQuantity'})

In [210]:
begInv = begInv.rename(columns = {'InvCost01_01_2016':'BeginningCost'})

In [211]:
endInv = endInv.rename(columns = {'InvCost12_31_2016':'EndingCost'})

In [212]:
deliver3=pd.merge(left=begInv,right=endInv,how="outer",on=["InventoryId"])

In [213]:
inventory=deliver3[["InventoryId","BeginningQuantity","BeginningCost","EndingQuantity","EndingCost"]]

In [214]:
inventory=inventory.fillna(0)

In [215]:
len(inventory)

256042

In [216]:
inventory

InventoryId  BeginningQuantity  BeginningCost  EndingQuantity  \
0       1_HARDERSFIELD_58               8.00         103.92           11.00   
1       1_HARDERSFIELD_60               7.00          76.93            0.00   
2       1_HARDERSFIELD_62               6.00         221.94            7.00   
3       1_HARDERSFIELD_63               3.00         116.97            7.00   
4       1_HARDERSFIELD_72               6.00         209.94            4.00   
...                   ...                ...            ...             ...   
256037  81_PEMBROKE_90087               0.00           0.00            3.00   
256038  81_PEMBROKE_90088               0.00           0.00            3.00   
256039  81_PEMBROKE_90089               0.00           0.00            3.00   
256040  81_PEMBROKE_90090               0.00           0.00            3.00   
256041  81_PEMBROKE_90604               0.00           0.00            2.00   

        EndingCost  
0           142.89  
1             0.00  
2           258.93  
3           272.93  
4           139.96  
...            ...  
256037    1,409.97  
256038      404.97  
256039      359.97  
256040    1,949.97  
256041      239.98  

[256042 rows x 5 columns]

## Deliverable 4A

In [217]:
purchases = pd.read_csv("PurchasesFINAL12-31-16Sample.csv")

In [218]:
purchases.head(1)

InventoryId  Store  Brand                Description   Size  \
0  69_MOUNTMEND_8412     69   8412  Tequila Ocho Plata Fresno  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2015-12-21   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2016-01-02  2016-01-04  2016-02-16          35.71         6   214.26   

   Classification  
0               1

In [219]:
groupTopVendors=purchases.groupby(["VendorNumber","VendorName"]).agg({"Dollars":"sum"})

In [220]:
groupTopVendors.head(1)

Dollars
VendorNumber VendorName                                      
2            IRA GOLDMAN AND WILLIAMS, LLP           1,751.30

In [221]:
sortedTopVendors=groupTopVendors.sort_values(by="Dollars",ascending=False)

In [222]:
sortedTopVendors.head(10)

Dollars
VendorNumber VendorName                               
3960         DIAGEO NORTH AMERICA INC    21,315,299.75
4425         MARTIGNETTI COMPANIES       11,709,222.19
12546        JIM BEAM BRANDS COMPANY     10,232,521.65
17035        PERNOD RICARD USA           10,224,899.93
480          BACARDI USA INC              7,420,630.63
1392         CONSTELLATION BRANDS INC     6,584,032.72
1128         BROWN-FORMAN CORP            5,719,375.62
9165         ULTRA BEVERAGE COMPANY LLP   5,545,766.29
3252         E & J GALLO WINERY           5,177,067.61
9552         M S WALKER INC               4,647,908.17

## Deliverable 4B

In [223]:
invoicepurchases = pd.read_csv("InvoicePurchases12-31-16Sample.csv")

In [224]:
invoicepurchases.head(1)

VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2016-01-04      8124   

       PODate     PayDate  Quantity  Dollars  Freight Approval  
0  2015-12-21  2016-02-16         6   214.26     3.47      NaN

In [225]:
groupVendorsFreight=invoicepurchases.groupby(["VendorNumber","VendorName"]).agg({"Freight":"sum"})

In [226]:
groupVendorsFreight.head(1)

Freight
VendorNumber VendorName                                      
2            IRA GOLDMAN AND WILLIAMS, LLP              27.08

In [227]:
sortedVendorsFrieght=groupVendorsFreight.sort_values(by="Freight",ascending=False)

In [228]:
sortedVendorsFrieght.head(10)

Freight
VendorNumber VendorName                            
3960         DIAGEO NORTH AMERICA INC    257,032.07
4425         MARTIGNETTI COMPANIES       144,719.92
12546        JIM BEAM BRANDS COMPANY     123,880.97
17035        PERNOD RICARD USA           123,780.22
480          BACARDI USA INC              89,286.27
1392         CONSTELLATION BRANDS INC     79,528.99
1128         BROWN-FORMAN CORP            68,601.68
9165         ULTRA BEVERAGE COMPANY LLP   68,054.70
3252         E & J GALLO WINERY           61,966.91
9552         M S WALKER INC               55,551.82

## Deliverable 4C

In [229]:
invoicepurchases["pfpd"]=((invoicepurchases["Freight"]/invoicepurchases["Dollars"])*100)

In [230]:
invoicepurchases.head(1)

VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2016-01-04      8124   

       PODate     PayDate  Quantity  Dollars  Freight Approval  pfpd  
0  2015-12-21  2016-02-16         6   214.26     3.47      NaN  1.62

In [231]:
freightpurchase=pd.merge(left=sortedTopVendors, right=sortedVendorsFrieght, how="inner", on=["VendorName"])

In [232]:
freightpurchase.head(1)

Dollars    Freight
VendorName                                          
DIAGEO NORTH AMERICA INC    21,315,299.75 257,032.07

In [233]:
above250k=freightpurchase[freightpurchase["Dollars"] > 250000]

In [234]:
above250k.head(1)

Dollars    Freight
VendorName                                          
DIAGEO NORTH AMERICA INC    21,315,299.75 257,032.07

In [235]:
above250k["PercentFreightPerDollar"]=((above250k["Freight"]/above250k["Dollars"])*100)

/var/folders/td/w649pg8d1bb2f3z2nxg773mr0000gn/T/ipykernel_281/4204502307.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  above250k["PercentFreightPerDollar"]=((above250k["Freight"]/above250k["Dollars"])*100)


In [236]:
above250k.head(1)

Dollars    Freight  PercentFreightPerDollar
VendorName                                                                   
DIAGEO NORTH AMERICA INC    21,315,299.75 257,032.07                     1.21

In [237]:
highestfreightcost=above250k.sort_values(by="PercentFreightPerDollar",ascending=False)

In [238]:
highestfreightcost.head(10)

Dollars    Freight  PercentFreightPerDollar
VendorName                                                                   
DIAGEO CHATEAU ESTATE WINES    573,875.03   7,259.75                     1.27
KOBRAND CORPORATION            733,971.86   9,139.38                     1.25
MARTIGNETTI COMPANIES       11,709,222.19 144,719.92                     1.24
STATE WINE & SPIRITS           650,999.58   8,014.98                     1.23
BANFI PRODUCTS CORP            692,669.85   8,510.41                     1.23
ULTRA BEVERAGE COMPANY LLP   5,545,766.29  68,054.70                     1.23
DELICATO VINEYARDS INC         652,181.26   7,982.65                     1.22
WINE GROUP INC               2,214,435.68  27,100.41                     1.22
SOUTHERN WINE & SPIRITS NE   1,541,712.12  18,808.83                     1.22
PROXIMO SPIRITS INC.         3,198,428.06  38,994.78                     1.22

In [239]:
lowestfreightcost=above250k.sort_values(by="PercentFreightPerDollar",ascending=True)

In [240]:
lowestfreightcost.head(5)

Dollars   Freight  PercentFreightPerDollar
VendorName                                                                 
SIDNEY FRANK IMPORTING CO     743,030.40  8,549.55                     1.15
TREASURY WINE ESTATES       1,269,997.73 14,836.57                     1.17
LUXCO INC                     878,296.35 10,261.60                     1.17
DISARONNO INTERNATIONAL LLC   498,383.55  5,861.03                     1.18
PALM BAY INTERNATIONAL INC  1,151,691.22 13,605.41                     1.18

## Deliverable 4D

In [241]:
purchases.head(1)

InventoryId  Store  Brand                Description   Size  \
0  69_MOUNTMEND_8412     69   8412  Tequila Ocho Plata Fresno  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2015-12-21   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2016-01-02  2016-01-04  2016-02-16          35.71         6   214.26   

   Classification  
0               1

In [242]:
itemspurchase=purchases.groupby(["VendorNumber","VendorName"]).agg({"Dollars":"sum","Quantity":"sum"})

In [243]:
itemspurchase.head(1)

Dollars  Quantity
VendorNumber VendorName                                                
2            IRA GOLDMAN AND WILLIAMS, LLP           1,751.30       101

In [244]:
itemspurchasefreight=pd.merge(left=itemspurchase, right=freightpurchase, how="outer", on=["VendorName","Dollars"])

In [245]:
itemspurchasefreight["FreightPerDollar2"]=(itemspurchasefreight["Freight"]/itemspurchasefreight["Dollars"])

In [246]:
itemspurchasefreight["PercentFreightPerUnit"]=((itemspurchasefreight["Freight"]/itemspurchasefreight["Quantity"])*100)

In [247]:
itemspurchasefreight["FreightPerUnit"]=(itemspurchasefreight["Freight"]/itemspurchasefreight["Quantity"])

In [248]:
itemspurchasefreight.head(1)

Dollars  Quantity  Freight  \
VendorName                                                            
IRA GOLDMAN AND WILLIAMS, LLP           1,751.30       101    27.08   

                                         FreightPerDollar2  \
VendorName                                                   
IRA GOLDMAN AND WILLIAMS, LLP                         0.02   

                                         PercentFreightPerUnit  FreightPerUnit  
VendorName                                                                      
IRA GOLDMAN AND WILLIAMS, LLP                            26.81            0.27

In [249]:
deliver4d1=itemspurchasefreight[itemspurchasefreight["Freight"] > 100]

In [250]:
deliver4d2=deliver4d1[deliver4d1["Quantity"] <= 1000]

In [251]:
deliver4d2sorted=deliver4d2.sort_values(by="Freight",ascending=False)

Top ten highest freight costs under the given conditions:

In [252]:
deliver4d2sorted.head(10)

Dollars  Quantity  Freight  \
VendorName                                                     
ATLANTIC IMPORTING COMPANY      20,321.72       903   211.74   
TAMWORTH DISTILLING             12,455.17       587   202.50   
CANDIA VINEYARDS                13,726.07       984   194.13   
SWEETWATER FARM                 15,557.89       726   178.34   
ALISA CARR BEVERAGES            16,935.18       170   172.00   
STARK BREWING COMPANY           10,581.48       494   130.09   
HOOD RIVER DISTILLERS, Inc.     15,277.56       723   117.52   

                                 FreightPerDollar2  PercentFreightPerUnit  \
VendorName                                                                  
ATLANTIC IMPORTING COMPANY                    0.01                  23.45   
TAMWORTH DISTILLING                           0.02                  34.50   
CANDIA VINEYARDS                              0.01                  19.73   
SWEETWATER FARM                               0.01                  24.56   
ALISA CARR BEVERAGES                          0.01                 101.18   
STARK BREWING COMPANY                         0.01                  26.33   
HOOD RIVER DISTILLERS, Inc.                   0.01                  16.25   

                                 FreightPerUnit  
VendorName                                       
ATLANTIC IMPORTING COMPANY                 0.23  
TAMWORTH DISTILLING                        0.34  
CANDIA VINEYARDS                           0.20  
SWEETWATER FARM                            0.25  
ALISA CARR BEVERAGES                       1.01  
STARK BREWING COMPANY                      0.26  
HOOD RIVER DISTILLERS, Inc.                0.16

## Deliverable 4E

The freight cost per dollar and per unit are included in "deliver4d2sorted" above. The percent freight per dollar is similar or slightly higher under the conditions specified for Deliverable 4D. The percent freight per unit is not very consistent, ranging from 16.25% to 101.18% among the ten vendors with the highest freight costs. 

## Deliverable 5

I decided to compare the number of items sold to the amount of revenue generated by each item. A business may not sell many of one item, that item may still be generating revenue. A business could use this information to determine which products it will keep and which products it will discontinue.

In [253]:
sales = pd.read_csv("salesDecember2016Sample.csv")

In [254]:
sales.head(1)

InventoryId  Store  Brand                   Description   Size  \
0  10_HORNSEY_1003     10   1003  Crown Royal +2 Reusable Cups  750mL   

   SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0              5        114.95       22.99  2016-12-01  750.00   

   Classification  ExciseTax  VendorNo                VendorName  
0               1       3.94      3960  DIAGEO NORTH AMERICA INC

Five items with the highest number of items sold in December 2016:

In [255]:
topsalesquant=sales.groupby(["Description"]).agg({"SalesQuantity":"sum"})

In [256]:
topsalesquantsort=topsalesquant.sort_values(by="SalesQuantity",ascending=False)

In [257]:
topsalesquantsort.head(5)

SalesQuantity
Description                          
Capt Morgan Spiced Rum          41450
Baileys Irish Cream             35525
Tito's Handmade Vodka           31855
Smirnoff 80 Proof               30118
Jameson Irish Whiskey           27636

Five items with the number of items sold in December 2016:

In [258]:
lowsalesquantsort=topsalesquant.sort_values(by="SalesQuantity",ascending=True)

In [259]:
lowsalesquantsort.head(5)

SalesQuantity
Description                                
von Buhl Jazz Rsl                         1
Guigal Cote Rotie Br et Blnd              1
Gruet Bl de Bl 25th Annivrsy              1
Grattamacco Bolgheri Rosso                1
Grape & Vine Merlot                       1

Five items with the highest dollar sales in December 2016:

In [260]:
salesdollars=sales.groupby(["Description"]).agg({"SalesDollars":"sum"})

In [261]:
salesdollarssort=salesdollars.sort_values(by="SalesDollars",ascending=False)

In [262]:
salesdollarssort.head(5)

SalesDollars
Description                          
Jameson Irish Whiskey      715,041.14
Capt Morgan Spiced Rum     631,362.50
Tito's Handmade Vodka      612,141.95
Jack Daniels No 7 Black    557,486.97
Baileys Irish Cream        555,702.75

Five items with the lowest dollar sales in December 2016:

In [263]:
lowsalesdollars=salesdollars.sort_values(by="SalesDollars",ascending=True)

In [264]:
lowsalesdollars.head(5)

SalesDollars
Description                               
Mojoshot Blue Lagoon RTD              0.99
UV Ruby Red Grapefruit Vodka          0.99
Shady Fruit Real Ruby Red Gr          1.98
Grape & Vine Merlot                   3.29
Liberty Creek Moscato                 3.49